.github/workflows/jenkins-security-scan.yml

```diff
@@ -0,0 +1,21 @@
+name: Jenkins Security Scan
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    types: [ opened, synchronize, reopened ]
+  workflow_dispatch:
+
+permissions:
+  security-events: write
+  contents: read
+  actions: read
+
+jobs:
+  security-scan:
+    uses: jenkins-infra/jenkins-security-scan/.github/workflows/jenkins-security-scan.yaml@v2
+    with:
+      java-cache: 'maven' # Optionally enable use of a build dependency cache. Specify 'maven' or 'gradle' as appropriate.
+      java-version: 11 # What version of Java to set up for the build.
```

Comment:
@strangelookingnerd why not 17 (or 21) these days?

Reply:
The scan did not run with the default so I downgraded the Java version.

Reply:
> The scan did not run with the default

Why? Is that normal? We expect everything to be built with 17 or 21 these days and may soon drop 11 support in plugins, so this is concerning.

Reply:
See https://github.com/jenkinsci/workflow-cps-plugin/pull/901/checks?check_run_id=26471237654 - seems to be related to `groovy-cps-dgm-builder`

Reply:
FTR

```
java.lang.reflect.InaccessibleObjectException: Unable to make public java.net.URI com.sun.tools.javac.file.PathFileObject$JarFileObject.toUri() accessible: module jdk.compiler does not "opens com.sun.tools.javac.file" to unnamed module @22d8cfe0
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at com.semmle.extractor.java.interceptors.JavacToolInterceptor.getTask(JavacToolInterceptor.java:196)
```

In general the autobuild is not great because we have no way of controlling how it works. For example we cannot tell it to https://github.com/jenkinsci/workflow-cps-plugin/blob/3f4affab2008b27b5666d7ef9cce0ea3cce485b9/lib/pom.xml#L61-L67

CC @daniel-beck @dwnusbaum

Reply:
> In general the autobuild is not great because we have no way of controlling how it works. For example we cannot tell it to 

Docs for use of Jenkins Security Scan rules with default CodeQL scan workflows are at https://github.com/jenkins-infra/jenkins-codeql?tab=readme-ov-file#use-in-a-regular-codeql-workflow

Reply:
> In general the autobuild is not great because we have no way of controlling how it works. For example we cannot tell it to

I don't really understand your point, did you mean "tell it to [ignore]"?

> at com.semmle.extractor.java.interceptors.JavacToolInterceptor.getTask(JavacToolInterceptor.java:196)

This is a CodeQL-specific class, which seems unexpected. I guess there is either a bug in CodeQL or our code. https://github.com/github/codeql/issues/7535 and https://github.com/github/codeql-action/issues/972 seem relevant. I couldn't find the relevant CodeQL source code though, maybe the actual important parts are closed source?  Here is the full stack trace, everything related to `semmle` is unusual.

```
  [2024-06-20 14:11:44] [autobuild] Exception in thread "main" java.lang.RuntimeException: Unknown file object: JarFileObject[/home/runner/work/workflow-cps-plugin/workflow-cps-plugin/lib/target/groovy-cps-dgm-builder-999999-SNAPSHOT-jar-with-dependencies.jar:/org/codehaus/groovy/runtime/StringGroovyMethods.java]
  [2024-06-20 14:11:44] [autobuild] 	at com.semmle.extractor.java.interceptors.JavacToolInterceptor.getTask(JavacToolInterceptor.java:216)
  [2024-06-20 14:11:44] [autobuild] 	at jdk.compiler/com.sun.tools.javac.api.JavacTool.SEMMLE_INTERCEPT$1(JavacTool.java)
  [2024-06-20 14:11:44] [autobuild] 	at jdk.compiler/com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:200)
  [2024-06-20 14:11:44] [autobuild] 	at jdk.compiler/com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:119)
  [2024-06-20 14:11:44] [autobuild] 	at jdk.compiler/com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:68)
  [2024-06-20 14:11:44] [autobuild] 	at com.cloudbees.groovy.cps.tool.Driver.run(Driver.java:59)
  [2024-06-20 14:11:44] [autobuild] 	at com.cloudbees.groovy.cps.tool.Driver.main(Driver.java:23)
  [2024-06-20 14:11:44] [autobuild] Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make public java.net.URI com.sun.tools.javac.file.PathFileObject$JarFileObject.toUri() accessible: module jdk.compiler does not "opens com.sun.tools.javac.file" to unnamed module @22d8cfe0
  [2024-06-20 14:11:44] [autobuild] 	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
  [2024-06-20 14:11:44] [autobuild] 	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
  [2024-06-20 14:11:44] [autobuild] 	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
  [2024-06-20 14:11:44] [autobuild] 	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
  [2024-06-20 14:11:44] [autobuild] 	at com.semmle.extractor.java.interceptors.JavacToolInterceptor.getTask(JavacToolInterceptor.java:196)
  [2024-06-20 14:11:44] [autobuild] 	... 6 more
```

In general, I don't really see why `com.semmle.extractor.java.interceptors.JavacToolInterceptor.getTask` is even trying to change the accessibility of `PathFileObject$JarFileObject.toUri()` with reflection. The class implements the [`JavaFileObject`](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/tools/JavaFileObject.html) API class that is part of the `java.compiler` module, which in turn implements [`FileObject`](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/tools/FileObject.html), and `FileObject.toUri` is part of the public API. Maybe it was originally targeting JDK 8 or something and needs to be updated.

Reply:
> did you mean "tell it to [ignore]"?

No, I meant that the `lib` module needs to compile against sources generated by a tool defined in the `dgm-builder` module, which is an unusual setup that an automatic builder might not be able to handle, but I guess it really runs Maven?

> or our code

https://github.com/jenkinsci/workflow-cps-plugin/blob/a6f67a8c907b4f3699bd54b46271e0ecd3caccc4/dgm-builder/src/main/java/com/cloudbees/groovy/cps/tool/Driver.java#L45 seems OK. AFAICT they somehow register a custom `javac` tool, and it is not fully compatible with Java 17? Maybe https://github.com/jenkinsci/workflow-cps-plugin/blob/a6f67a8c907b4f3699bd54b46271e0ecd3caccc4/dgm-builder/src/main/java/com/cloudbees/groovy/cps/tool/Driver.java#L27 needs to explicitly reject the Semmle/CodeQL implementation. (After all, we only need CodeQL to run on the `plugin` module.)

Indeed the core CodeQL source code seems to be private.

Reply:
> I guess it really runs Maven?

Yes, it is running `mvn clean package -f pom.xml -B -V -e -Dfindbugs.skip -Dcheckstyle.skip -Dpmd.skip=true -Dspotbugs.skip -Denforcer.skip -Dmaven.javadoc.skip -DskipTests -Dmaven.test.skip.exec -Dlicense.skip=true -Drat.skip=true -Dspotless.check.skip=true -t /home/runner/.m2/toolchains.xml`, which seems more or less fine.

> needs to explicitly reject the Semmle/CodeQL implementation

Let's see if https://github.com/jenkinsci/workflow-cps-plugin/pull/917 works.